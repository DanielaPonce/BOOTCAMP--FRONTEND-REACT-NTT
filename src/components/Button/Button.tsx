import { FC } from 'react';

interface ButtonProps {
	type?: 'button' | 'submit' | 'reset';
	onClick?: () => void;
	disabled?: boolean;
	children: React.ReactNode;
	className?: string;
}

const Button: FC<ButtonProps> = ({
	type = 'button',
	onClick,
	disabled = false,
	children,
	className = ''
}) => {
	return (
		<button
			type={type}
			onClick={onClick}
			disabled={disabled}
			className={`btn ${className} ${disabled ? 'btn-disabled' : ''}`}
		>
			{children}
		</button>
	);
};

export default Button;
