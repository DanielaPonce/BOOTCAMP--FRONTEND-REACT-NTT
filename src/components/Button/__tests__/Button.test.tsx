import { fireEvent, render, screen } from '@testing-library/react';
import Button from '../Button';

describe('Button Component', () => {
	test('renders correctly with default props', () => {
		const text = 'Add to cart';
		render(<Button>{text}</Button>);
		const button = screen.getByRole('button');
		expect(button).toBeInTheDocument();
		expect(button).toHaveTextContent(text);
		expect(button).toHaveClass('btn');
		expect(button).not.toBeDisabled();
	});

	test('applies the className prop correctly', () => {
		render(<Button className="custom-class">Add to cart</Button>);
		const button = screen.getByRole('button');
		expect(button).toHaveClass('custom-class');
	});

	test('renders with the correct type attribute', () => {
		render(<Button type="submit">Submit</Button>);
		const button = screen.getByRole('button');
		expect(button).toHaveAttribute('type', 'submit');
	});

	test('disables the button when the disabled prop is true', () => {
		render(<Button disabled={true}>Disabled</Button>);
		const button = screen.getByRole('button');
		expect(button).toBeDisabled();
		expect(button).toHaveClass('btn-disabled');
	});

	test('calls onClick when the button is clicked', () => {
		const handleClick = jest.fn();
		render(<Button onClick={handleClick}>Click Me</Button>);
		const button = screen.getByRole('button');
		fireEvent.click(button);
		expect(handleClick).toHaveBeenCalledTimes(1);
	});

	test('does not call onClick when the button is disabled', () => {
		const handleClick = jest.fn();
		render(
			<Button onClick={handleClick} disabled={true}>
				Add to cart
			</Button>
		);
		const button = screen.getByRole('button');
		fireEvent.click(button);
		expect(handleClick).not.toHaveBeenCalled();
	});

	test('renders children correctly', () => {
		render(
			<Button>
				<span data-testid="child">Child Content</span>
			</Button>
		);
		const child = screen.getByTestId('child');
		expect(child).toBeInTheDocument();
		expect(child).toHaveTextContent('Child Content');
	});

	test('matches snapshot', () => {
		const { asFragment } = render(
			<Button className="custom-class" disabled={true}>
				Snapshot Button
			</Button>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
