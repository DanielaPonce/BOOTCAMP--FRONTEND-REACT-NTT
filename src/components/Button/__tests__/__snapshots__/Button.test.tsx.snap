// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn custom-class btn-disabled"
    disabled=""
    type="button"
  >
    Snapshot Button
  </button>
</DocumentFragment>
`;
